### ISSUE TITLE
Puback iscalled too soon on incoming publish message

### ISSUE THREAD

**Issue by tomasol** [AtherEnergy/rumqtt#94]
With QoS 1 I expect that when my app crashes while processing incoming message, after my app starts again I will receive the message again. At least that's how I understand "at least once" delivery.
After my testing I looked at the code: [connection.rs#L546](https://github.com/AtherEnergy/rumqtt/blob/08c2e892b575c0519edcfaf317e596064b10819c/src/connection.rs#L546) and realized this is not the case. Puback is sent right after publish packet is received, so there is no chance the app can persist and not lose the packet in case of crash. I believe this is wrong.
One way would be to call `puback(pkid)` only after `on_message` handler finishes.
Other way would be to give user a way to call `puback(pkid)` explicitly.

**Comment by tekjar** [comment]
@tomasol When Mqtt protocol says At least once, that doesn't mean mqtt broker has to hold the message until client is done processing the message. I just means that broker has to hold the message until client receives the message. This helps the client to be robust in bad networks. At least that's what I understood. 

I feel there is a reason why mqtt is this way

* What if your code always panics while handling the message? Broker has to hold the message forever because the user code is wrong?

**Comment by tekjar** [comment]
Maybe it's debatable why not but it'll be a different protocol. Spec clearly says `message arrives at the receiver`

http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718101

**Comment by tomasol** [comment]
Thanks for response. 
> What if your code always panics while handling the message? Broker has to hold the message forever because the user code is wrong?

There is already a window of opportunity for that, e.g. if processing incoming pub packet fails (I mean in any library, not yours in particular). QoS cannot protect you from bugs, so I think that is out of scope. 

It should however protect from losing messages due to power/network outage. If the outage happens before puback is sent & received, broker just resends the message. If however the outage happens while I try to persist the data, I may never be able to recover it. There are applications where it does not matter, but most applications would not work correctly under these circumstances.
Spec says:

> In the QoS 1 delivery protocol, the Receiver
> MUST respond with a PUBACK Packet containing the Packet Identifier from the incoming PUBLISH Packet, having accepted ownership of the Application Message

IMO this is clear rerefence of owning the data and being responsible for persisting it if needed.

It is even more obvious for QoS2 but let's not get into that. Also, there are mqtt libraries that have persistence for QoS right in their design, e.g. [paho java client](https://www.eclipse.org/paho/files/javadoc/org/eclipse/paho/client/mqttv3/MqttClientPersistence.html)

**Comment by tomasol** [comment]
Thank you for your response. We can probably agree that the spec leaves some room for interpretation, ultimately it is the application's responsibility to decide what's the correct behavior for its use case.

> I feel it's the other way around. Lot of client applications will be okay to loose data during power outages/crashes but for some it's not an option

It's hard to say what percentage of apps tolerate data loss, I just think it would be nice to have the ability to decide when the ack should be sent. Some libraries support persistence, some don't. I like the way the java client handles it - not only is there the mentioned persistence api, but you can use a [callback](https://www.eclipse.org/paho/files/javadoc/index.html?org/eclipse/paho/client/mqttv3/MqttCallback.html) that gets called before `puback` is sent. Also it is possible to turn on [manual acks](https://www.eclipse.org/paho/files/javadoc/org/eclipse/paho/client/mqttv3/IMqttClient.html#setManualAcks-boolean-) for even finer control.

Either one of those options would be sufficient and would give apps more robustness and protect them from data loss.

**Comment by tekjar** [comment]
@tomasol Agreed. I'll leave this issue open and try to cater this in the new design. Maybe not immediately but eventually I'll add the support to send acks after handling
